feat(couchdb-client): add createDb method and refactor CouchdbClient implementation

jmgang opened the pull request:
This PR introduces a new feature in the CouchdbClient, adding the `createDb` method to allow database creation with optional parameters. Additionally, the CouchdbClient implementation has been refactored to include a new options class for easier configuration when creating the client. 

Let me know your thoughts on this PR @Stwissel and @rjtmahinay.

Stwissel commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -5,74 +5,67 @@
-    if (this.credentials != null) {
-      request.authentication(this.credentials);
+    if (clientOptions.getCredentials() != null) {
+      request.authentication(clientOptions.getCredentials().toAuthCredentials());
```
We can write that smarter:

`public class AnonymousCrerdentials implements Credentials`

see: https://vertx.io/docs/apidocs/io/vertx/ext/auth/authentication/Credentials.html

then 

```java
Credentials getCredentials() {
return clientOptions.getCredentials() != null ? clientOptions.getCredentials() : new AnonymousCrerdentials();
}

```

jmgang replied:
Got it. I'll check and refactor it

Stwissel commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -81,26 +74,57 @@ public Future<Buffer> rawCall(JsonObject params) {
+    if (clientOptions.getCredentials() != null) {
+      request.authentication(clientOptions.getCredentials().toAuthCredentials());
     }
     request.send()
```
need at some time account for streaming bodies here

jmgang replied:
sure, but it might not be included in this PR yet

Stwissel commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -81,26 +74,57 @@ public Future<Buffer> rawCall(JsonObject params) {
   private void initClient() {
     if (client == null) {
-      client = WebClient.create(vertx, options);
+      client = WebClient.create(vertx, clientOptions.getWebClientOptions());
```
Better as dependency injection

Stwissel commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -81,26 +74,57 @@ public Future<Buffer> rawCall(JsonObject params) {
-  public CouchdbClient authetication(Credentials credentials) {
-    this.credentials = credentials;
-    return this;
+  public Future<JsonObject> createDb(JsonObject options) {
```
```
public Future<JsonObject> createDb(String databaseName)
public Future<JsonObject> createDb(String databaseName, JsonObject options)
```

removes the need to create a JsonObject for the default case

Stwissel commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -81,26 +74,57 @@ public Future<Buffer> rawCall(JsonObject params) {
+        if (statusCode == 201 || statusCode == 202) {
+          promise.complete(result);
+        } else {
+          promise.fail(new RuntimeException(
```
We might want our own exception that can carry the returned body

jmgang replied:
Noted

Stwissel commented on src/test/java/io/vertx/ext/couchdb/CouchdbClientIT.java:
```diff
@@ -0,0 +1,139 @@
+
+  @BeforeAll
+  static void setup() {
+    vertx = Vertx.vertx();
```
There's a JUnit Vert.x extension that allows to inject vertx:
https://vertx.io/docs/vertx-junit5/java/#_integration_with_junit_5

Stwissel commented on src/test/java/io/vertx/ext/couchdb/CouchdbClientTest.java:
```diff
@@ -1,31 +1,58 @@
 import io.vertx.core.json.JsonArray;
 import io.vertx.core.json.JsonObject;
-import io.vertx.ext.auth.authentication.UsernamePasswordCredentials;
+import io.vertx.ext.couchdb.impl.CouchdbClientImpl;
```
better `io.vertx.ext.couchdb.CouchdbClient;`

Stwissel commented on src/test/java/io/vertx/ext/couchdb/CouchdbClientTest.java:
```diff
@@ -1,31 +1,58 @@
-    client = CouchdbClient.create(vertx, options);
-    client.authetication(new UsernamePasswordCredentials("admin", "password"));
+  void setUp() throws Exception {
+    Vertx vertx = Vertx.vertx();
```
Use `@ExtendWith(VertxExtension.class)` -> add to the `@IntregrationTest` annotation

jmgang replied:
Already added this at `@IntegrationTest`annotation. Would just need to apply it.

Stwissel commented on src/test/java/io/vertx/ext/couchdb/utils/IntegrationTest.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * IntegrationTest annotation
+ */
+package io.vertx.ext.couchdb.utils;
```
`io.vertx.ext.couchdb.testannotations;` ?

jmgang replied:
Sure, that would be a better package name

Stwissel reviewed (commented):
A good start, lets discuss

jmgang commented on src/main/java/io/vertx/ext/couchdb/CouchdbClientOptions.java:
```diff
@@ -0,0 +1,107 @@
+    return database;
+  }
+
+  public CouchdbClientOptions setDatabase(String database) {
```
I guess not. I'll refactor this

jmgang commented on src/main/java/io/vertx/ext/couchdb/CouchdbClientOptions.java:
```diff
@@ -0,0 +1,107 @@
+    return this;
+  }
+
+  public CouchdbCredentials getCredentials() {
```
Yes. The reason of why I added a new `CouchdbCredentials` was because I am using `@JsonGen` that can generate a `CouchdbClientOptionsConverter `class which can be used for converting from and to json like this:

```
public CouchdbClientOptions(JsonObject json) {
    this();
    CouchdbClientOptionsConverter.fromJson(json, this);
  }
```
However the `CouchdbClientOptionsConverter `class expects all the fields of the `CouchdbClientOptions` to be a concrete class thus I cannot use the `io.vertx.ext.auth.authentication.Credentials` class because it's an interface. But please let me know if you have any ideas around this or if we still want to generate a converter class in the first place.

Stwissel replied:
mark it as transient? Or redesign the options not to have credentials. It's important to stay with the interface here

jmgang replied:
I would lean towards with redesigning the options. Maybe we can just remove the options class and simply use a builder pattern for the `CouchdbClient ` for simplicity and accessibility

Stwissel replied:
retire CouchDbClientOptions for webclient in constructor

jmgang commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -5,74 +5,67 @@
 import io.vertx.core.Vertx;
 import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpMethod;
+import io.vertx.core.internal.VertxInternal;
```
alright I'll revert this back

jmgang commented on src/main/java/io/vertx/ext/couchdb/CouchdbCredentials.java:
```diff
@@ -0,0 +1,75 @@
+@DataObject
+@JsonGen
+public class CouchdbCredentials {
+
```
Yes. The reason of why I added a new `CouchdbCredentials` was because I am using `@JsonGen` that can generate a `CouchdbClientOptionsConverter `class which can be used for converting from and to json like this:

```
public CouchdbClientOptions(JsonObject json) {
    this();
    CouchdbClientOptionsConverter.fromJson(json, this);
  }
```
However the `CouchdbClientOptionsConverter `class expects all the fields of the `CouchdbClientOptions` to be a concrete class thus I cannot use the `io.vertx.ext.auth.authentication.Credentials` because it's an interface. But please let me know if you have any ideas around this or if we still want to generate a converter class in the first place.

Stwissel replied:
would transient work? Also interesting question: What options beyond the webclient options do we want to introduce?

jmgang commented on src/main/java/io/vertx/ext/couchdb/impl/CouchdbClientImpl.java:
```diff
@@ -5,74 +5,67 @@
-  private final WebClientOptions options;
+  private final VertxInternal vertx;
+  private final CouchdbClientOptions clientOptions;
   private WebClient client = null;
```
was not aware of this. I'll look into this

jmgang commented on src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:
```diff
@@ -27,7 +25,7 @@ public interface CouchdbClient {
    * @return the client
    */
-  static CouchdbClient create(Vertx vertx, WebClientOptions options) {
+  static CouchdbClient create(Vertx vertx, CouchdbClientOptions options) {
```
I think adding `public` here would be redundant as `CouchdbClient` is an interface

Stwissel replied:
that doesn't make it public when assigned to a method

jmgang commented on src/main/java/io/vertx/ext/couchdb/CouchdbCredentials.java:
```diff
@@ -0,0 +1,75 @@
+    } else if (token != null) {
+      return new TokenCredentials(token);
+    } else {
+      throw new IllegalArgumentException("Invalid credentials: either username and password, or token must be provided.");
```
Update: I would use the `Credentials` interface so  I'll remove this custom credentials class and let the developer define what `Credentials `they would want to use

jmgang commented:
hi @Stwissel, I've already addressed the code review comments and implemented what we had discussed in the progress check meeting. Kindly take a look at the changes. Thanks

Stwissel commented:
PRs shouldn't survive midnight 😎
